extractor/docker/docker.go

```diff
@@ -186,41 +180,72 @@ func ApplyLayers(layers []types.LayerInfo) types.ImageDetail {
 }
 
 func (d Extractor) ImageName() string {
-	return d.image.Name()
+	return d.imageName
 }
 
-func (d Extractor) ImageID() digest.Digest {
-	return d.image.ConfigInfo().Digest
+func (d Extractor) ImageID() (string, error) {
+	h, err := d.image.ConfigName()
+	if err != nil {
+		return "", xerrors.Errorf("unable to get the image ID: %w", err)
+	}
+	return h.String(), nil
 }
 
-func (d Extractor) ConfigBlob(ctx context.Context) ([]byte, error) {
-	return d.image.ConfigBlob(ctx)
+func (d Extractor) ConfigBlob() ([]byte, error) {
+	return d.image.RawConfigFile()
 }
 
-func (d Extractor) LayerIDs() []string {
-	return d.image.LayerIDs()
+func (d Extractor) LayerIDs() ([]string, error) {
+	conf, err := d.image.ConfigFile()
+	if err != nil {
+		return nil, xerrors.Errorf("unable to get the config file: %w", err)
+	}
+
+	var layerIDs []string
+	for _, d := range conf.RootFS.DiffIDs {
+		layerIDs = append(layerIDs, d.String())
+	}
+	return layerIDs, nil
 }
 
-func (d Extractor) ExtractLayerFiles(ctx context.Context, dig digest.Digest, filenames []string) (
-	digest.Digest, extractor.FileMap, []string, []string, error) {
-	img, err := d.image.GetLayer(ctx, dig)
+func (d Extractor) ExtractLayerFiles(diffID string, filenames []string) (string, extractor.FileMap, []string, []string, error) {
+	// diffID is a hash of the uncompressed layer
+	h, err := v1.NewHash(diffID)
+	if err != nil {
+		return "", nil, nil, nil, xerrors.Errorf("invalid layer ID (%s): %w", diffID, err)
+	}
+
+	layer, err := d.image.LayerByDiffID(h)
 	if err != nil {
-		return "", nil, nil, nil, xerrors.Errorf("failed to get a blob: %w", err)
+		return "", nil, nil, nil, xerrors.Errorf("failed to get the layer (%s): %w", diffID, err)
 	}
-	defer img.Close()
 
-	// calculate decompressed layer ID
-	sha256hash := sha256.New()
-	r := io.TeeReader(img, sha256hash)
+	r, err := layer.Uncompressed()
+	if err != nil {
+		return "", nil, nil, nil, xerrors.Errorf("failed to get the layer content (%s): %w", diffID, err)
+	}
+
+	// digest is a hash of the compressed layer
+	var digest string
+	if d.isCompressed(layer) {
+		d, err := layer.Digest()
+		if err != nil {
+			return "", nil, nil, nil, xerrors.Errorf("failed to get the digest (%s): %w", diffID, err)
+		}
+		digest = d.String()
+	}
 
 	files, opqDirs, whFiles, err := d.extractFiles(r, filenames)
 	if err != nil {
 		return "", nil, nil, nil, xerrors.Errorf("failed to extract files: %w", err)
 	}
 
-	decompressedLayerID := digest.NewDigestFromBytes(digest.SHA256, sha256hash.Sum(nil))
+	return digest, files, opqDirs, whFiles, nil
+}
 
-	return decompressedLayerID, files, opqDirs, whFiles, nil
+func (d Extractor) isCompressed(l v1.Layer) bool {
+	_, uncompressed := reflect.TypeOf(l).Elem().FieldByName("UncompressedLayer")
```

Comment:
What if `l` is `nil`?

Reply:
If `l` gets to be nil, `LayerByDiffID` will return an error. It means `isCompressed` will not be called.
https://github.com/aquasecurity/fanal/blob/26afaa6f737011f7a821e7df4007cda69b5daaa4/extractor/docker/docker.go#L218-L221

Reply:
I understand.